### PR TITLE
build: Simpler Docker build

### PR DESCRIPTION
**Pull request by alexec**
| Build | Before | After |
|---|---|---|
| Docker Build | 1h | 45m | 
| CI | 13m | 17m |

### CODE REVIEW

File: .dockerignore [hunk 1]
```diff
@@ -1,10 +1,21 @@
 # Prevent vendor directory from being copied to ensure we are not not pulling unexpected cruft from
 # a user's workspace, and are only building off of what is locked by dep.
+*.iml
```
Comment (alexec) [inline .dockerignore:3]: while these files are small, they can result in invalidating the Docker caches, forcing build steps to be re-run

File: Dockerfile [hunk 2]
```diff
@@ -1,11 +1,11 @@
-####################################################################################################
-# Builder image
-# Initial stage which pulls prepares build dependencies and CLI tooling we need for our final image
-# Also used as the image in CI jobs so needs all dependencies
-####################################################################################################
-FROM golang:1.15.7 as builder
 
-ARG IMAGE_OS=linux
+ARG DOCKER_CHANNEL=stable
```
Comment (alexec) [inline Dockerfile:2]: args can and should go before the first FROM

File: Dockerfile [hunk 3]
```diff
@@ -27,32 +27,24 @@ RUN apt-get update && apt-get --no-install-recommends install -y \
...
-        if [ "${IMAGE_ARCH}" = "ppc64le" ] ||[ "${IMAGE_ARCH}" = "s390x" ]; then \
-	        wget -O docker.tgz https://download.docker.com/${IMAGE_OS}/static/${DOCKER_CHANNEL}/${IMAGE_ARCH}/docker-18.06.3-ce.tgz; \
-        else \
-            wget -O docker.tgz https://download.docker.com/${IMAGE_OS}/static/${DOCKER_CHANNEL}/${IMAGE_ARCH}/docker-${DOCKER_VERSION}.tgz; \
-        fi \
-    fi && \
-    tar --extract --file docker.tgz --strip-components 1 --directory /usr/local/bin/ && \
-    rm docker.tgz
+# https://blog.container-solutions.com/faster-builds-in-docker-with-go-1-11
+WORKDIR /go/src/github.com/argoproj/argo
```
Comment (alexec) [inline Dockerfile:31]: classic cache speed up technique

File: Dockerfile [hunk 4]
```diff
@@ -27,32 +27,24 @@ RUN apt-get update && apt-get --no-install-recommends install -y \
...
+COPY . .
 
 ####################################################################################################
-# argoexec-base
-# Used as the base for both the release and development version of argoexec
-####################################################################################################
+
 FROM debian:10.7-slim as argoexec-base
 
-ARG IMAGE_OS=linux
```
Comment (alexec) [inline Dockerfile:51]: we don't need this because we use `docker buildx --platform linux/amd64'

File: Dockerfile [hunk 5]
```diff
@@ -64,85 +56,89 @@ RUN apt-get update && \
...
+COPY hack/nsswitch.conf /etc/
+
 
 ####################################################################################################
 
 FROM node:14.0.0 as argo-ui
 
-ADD ["ui", "ui"]
-ADD ["api", "api"]
+COPY ui/package.json ui/yarn.lock ui/
```
Comment (alexec) [inline Dockerfile:80]: another classic speed up technique

File: Dockerfile [hunk 6]
```diff
@@ -64,85 +56,89 @@ RUN apt-get update && \
...
 
-ARG IMAGE_OS=linux
+FROM builder as workflow-controller-build
 
-# Perform the build
-WORKDIR /go/src/github.com/argoproj/argo
-COPY . .
-# check we can use Git
-RUN git rev-parse HEAD
+RUN --mount=type=cache,target=/root/.cache/go-build make dist/workflow-controller
```
Comment (alexec) [inline Dockerfile:99]: this is a buildx way to do build caching, I'm not 100% sure it works, so we may want to remove later on

File: Dockerfile [hunk 7]
```diff
@@ -64,85 +56,89 @@ RUN apt-get update && \
...
 USER 8737
-ARG IMAGE_OS=linux
-COPY --from=argoexec-base /etc/ssh/ssh_known_hosts /etc/ssh/ssh_known_hosts
-COPY --from=argoexec-base /etc/nsswitch.conf /etc/nsswitch.conf
-COPY --from=argoexec-base /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/ca-certificates.crt
-COPY --from=argo-build --chown=8737 /go/src/github.com/argoproj/argo/argo-server.crt argo-server.crt
-COPY --from=argo-build --chown=8737 /go/src/github.com/argoproj/argo/argo-server.key argo-server.key
-COPY --from=argo-build /go/src/github.com/argoproj/argo/dist/argo-${IMAGE_OS}-* /bin/argo
+
+COPY hack/ssh_known_hosts /etc/ssh/
```
Comment (alexec) [inline Dockerfile:137]: no longer depend on argoexec

File: Makefile [hunk 8]
```diff
@@ -16,39 +13,22 @@ GIT_TAG                = $(shell git describe --always --tags --abbrev=0 || echo
...
 
 VERSION               := latest
-DEV_IMAGE             := true
 DOCKER_PUSH           := false
 
 # VERSION is the version to be used for files in manifests and should always be latest uunlesswe are releasing
 # we assume HEAD means you are on a tag
 ifeq ($(findstring release,$(GIT_BRANCH)),release)
 VERSION               := $(GIT_TAG)
-DEV_IMAGE             := false
```
Comment (alexec) [inline Makefile:37]: no longer need dev images, standard build is fast enough

File: Makefile [hunk 9]
```diff
@@ -16,39 +13,22 @@ GIT_TAG                = $(shell git describe --always --tags --abbrev=0 || echo
...
-# If we are building dev images, then we want to use the Docker cache for speed.
-ifeq ($(DEV_IMAGE),true)
-DOCKERFILE            := Dockerfile.dev
-endif
-
-# version change, so does the file location
-CLI_IMAGE_FILE         := dist/cli-image.marker
-EXECUTOR_IMAGE_FILE    := dist/executor-image.marker
-CONTROLLER_IMAGE_FILE  := dist/controller-image.marker
+CGO_ENABLED           ?= 1
```
Comment (alexec) [inline Makefile:30]: TODO: delete

File: Dockerfile [hunk 10]
```diff
@@ -27,32 +27,24 @@ RUN apt-get update && apt-get --no-install-recommends install -y \
...
-    fi && \
-    tar --extract --file docker.tgz --strip-components 1 --directory /usr/local/bin/ && \
-    rm docker.tgz
+# https://blog.container-solutions.com/faster-builds-in-docker-with-go-1-11
+WORKDIR /go/src/github.com/argoproj/argo
+COPY go.mod .
+COPY go.sum .
+RUN go mod download
+
+WORKDIR /go/src/github.com/argoproj/argo
```
Comment (terrytangyuan) [inline Dockerfile:36]: argo -> argo-workflows

**Comment by alexec** [comment]
This is actually a 3m slower to run tests CI, but is faster to start-up locally.

**Review by simster7 — APPROVED** [review]

**Pull request merged**